File: ch4/mst/kruskal.java
import java.util.*;
import java.io.*;

// Union-Find Disjoint Sets Library written in OOP manner, using both path compression and union by rank heuristics
class UnionFind {                                              // OOP style
  private ArrayList<Integer> p, rank, setSize;
  private int numSets;

  public UnionFind(int N) {
    p = new ArrayList<Integer>(N);
    rank = new ArrayList<Integer>(N);
    setSize = new ArrayList<Integer>(N);
    numSets = N;
    for (int i = 0; i < N; i++) {
      p.add(i);
      rank.add(0);
      setSize.add(1);
    }
  }

  public int findSet(int i) { 
    if (p.get(i) == i) return i;
    else {
      int ret = findSet(p.get(i)); p.set(i, ret);
      return ret;
    }
  }

  public Boolean isSameSet(int i, int j) { return findSet(i) == findSet(j); }

  public void unionSet(int i, int j) { 
    if (!isSameSet(i, j)) { numSets--; 
    int x = findSet(i), y = findSet(j);
    // rank is used to keep the tree short
    if (rank.get(x) > rank.get(y)) { p.set(y, x); setSize.set(x, setSize.get(x) + setSize.get(y)); }
    else                           { p.set(x, y); setSize.set(y, setSize.get(y) + setSize.get(x));
                                     if (rank.get(x) == rank.get(y)) rank.set(y, rank.get(y) + 1); } } }
  public int numDisjointSets() { return numSets; }
  public int sizeOfSet(int i) { return setSize.get(findSet(i)); }
}

public class kruskal {
  public static void main(String[] args) throws Exception {
    /*
    // Graph in Figure 4.10 left, format: list of weighted edges
    // This example shows another form of reading graph input
    5 7
    0 1 4
    0 2 4
    0 3 6
    0 4 6
    1 2 2
    2 3 8
    3 4 9
    */

    File f = new File("mst_in.txt");
    Scanner sc = new Scanner(f);

    // Kruskal's algorithm
    int V = sc.nextInt(), E = sc.nextInt();
    ArrayList<IntegerTriple> EL = new ArrayList<>();
    for (int i = 0; i < E; ++i) {
      int u = sc.nextInt(), v = sc.nextInt(), w = sc.nextInt();
      EL.add(new IntegerTriple(w, u, v));        // reorder as (w, u, v)
    }
    Collections.sort(EL);                        // sort by w, O(E log E)

    int mst_cost = 0, num_taken = 0;             // no edge has been taken
    UnionFind UF = new UnionFind(V);             // all V are disjoint sets
    for (int i = 0; i < E; ++i) {                // up to O(E)
      IntegerTriple front = EL.get(i);
      if (UF.isSameSet(front.second(), front.third())) continue; // check
      mst_cost += front.first();                 // add w of this edge
      UF.unionSet(front.second(), front.third());// link them
      ++num_taken;                               // 1 more edge is taken
      if (num_taken == V-1) break;               // optimization
    }

    // note: the number of disjoint sets must eventually be 1 for a valid MST
    System.out.printf("MST cost = %d (Kruskal's)\n", mst_cost);
  }
}

File: ch4/mst/kruskal.ml
module UnionFind = struct
  type t =
    { parent: int array;
      rank: int array;
      size: int array;
      num_sets: int ref;
    }
  
  let create n =
    { parent  = Array.init n (fun x -> x);
      rank    = Array.make n 0;
      size    = Array.make n 1;
      num_sets = ref n;
    }
  
  let find { parent } u =
    let rec aux u =
      if parent.(u) = u then u
      else begin
        parent.(u) <- aux parent.(u);
        parent.(u)
      end
    in aux u
  
  let is_same_set dsu u v = find dsu u = find dsu v

  let union dsu u v =
    let link x p =
      dsu.parent.(x) <- p;
      dsu.size.(p) <- dsu.size.(p) + dsu.size.(x);
      if dsu.rank.(x) = dsu.rank.(p) then
        dsu.rank.(p) <- dsu.rank.(p) + 1
    in
    let pu, pv = find dsu u, find dsu v in
    if pu = pv then ()
    else begin
      decr dsu.num_sets;
      if dsu.rank.(pu) > dsu.rank.(pv) then link pv pu
      else link pu pv
    end
  
  let num_disjoint_sets { num_sets } = !num_sets

  let size_of_set dsu u = dsu.size.(find dsu u)
end

(** [kruskal n edges] returns the total weight of a MST of the input graph [(0..n-1, edges)] *)
let kruskal n edges =
  let dsu = UnionFind.create n in
  let cost = ref 0 in
  edges
    |> List.sort (fun (_, _, w1) (_, _, w2) -> compare w1 w2)
    |> List.iter (fun (u, v, w) ->
        if not (UnionFind.is_same_set dsu u v) then begin
          UnionFind.union dsu u v;
          cost := !cost + w
        end
      );
  !cost

open Scanf
open Printf

let sc = object method nf = bscanf Scanning.stdin end

let () =
  let (n, m) = sc#nf "%d %d\n" (fun x y -> (x, y)) in
  let edges = List.init m (fun _ -> sc#nf "%d %d %d\n" (fun u v w -> (u, v, w))) in
  kruskal n edges |> printf "MST cost = %d (Kruskal's)\n"
